src/parser.js

```diff
@@ -818,74 +800,94 @@ export default class Parser {
-        if (interpolations.length) {
-            hasId = hasId.filter(hashIndex => !~interpolations.indexOf(hashIndex));
+
+        // Get the content of each token
+        const tokensContent = tokensList.map(token => {
+            if (token[TOKEN.TYPE] === tokens.space) {
+                return this.requiredSpace(token);
+            }
+            return this.content(token);
+        });
+
+        // Parse the list of tokens and create a list of new nodes
+        const nodesToCreate = [];
+        let inProgressNode;
+        tokensList.forEach((token, tokenIndex) => {
+            const content = tokensContent[tokenIndex];
+            for (let i = 0; i < content.length; i++) {
+                const char = content[i];
+                const prevChar = content[i - 1] || (tokenIndex !== 0 ? tokensContent[tokenIndex - 1].slice(-1) : undefined);
+                const nextChar = content[i + 1] || (tokenIndex !== tokensContent.length - 1 ? tokensContent[tokenIndex + 1][0] : undefined);
+
+                if (char === "." && prevChar !== "\\") {
+                    initNode(ClassName);
+                } else if (char === "#" && prevChar !== "\\" && nextChar !== "{") {
+                    initNode(ID);
+                } else if (!inProgressNode) {
+                    initNode(Tag, char);
+                } else {
+                    inProgressNode.value += char;
+                    inProgressNode.endToken = token;
+                    inProgressNode.endIndex = i;
+                }
+
+                function initNode (NodeConstructor, value = "") {
+                    if (inProgressNode) {
+                        nodesToCreate.push(inProgressNode);
+                    }
+                    inProgressNode = {
+                        NodeConstructor,
+                        value,
+                        startToken: token,
+                        endToken: token,
+                        startIndex: i,
+                        endIndex: i,
+                    };
+                }
+            }
+        });
+        if (inProgressNode) {
+            nodesToCreate.push(inProgressNode);
         }
-        let indices = sortAsc(uniqs([0, ...hasClass, ...hasId]));
-        indices.forEach((ind, i) => {
-            const index = indices[i + 1] || word.length;
-            const value = word.slice(ind, index);
+
+        nodesToCreate.forEach((node, i) => {
             if (i === 0 && firstCallback) {
-                return firstCallback.call(this, value, indices.length);
+                firstCallback.call(this, node.value, nodesToCreate.length);
+                return;
             }
-            let node;
-            const current = this.currToken;
-            const sourceIndex = current[TOKEN.START_POS] + indices[i];
+            const {NodeConstructor, value, startToken, endToken, startIndex, endIndex} = node;
+            const sourceIndex = startToken[TOKEN.START_POS] + startIndex;
```

Comment:
This was basically the fix - we're now using `startToken` and `endToken` which are correctly computed for each class/id/tag, instead of always using `this.currToken`